File: Data-Structures-and-Algorithms-using-PHP/String/StringTree.php
<?php
class Node {
	public $value;
	public $count;
	public $lChild;
	public $rChild;
	Function __construct($value) {
		$this->value = $value;
		$this->lChild = null;
		$this->rChild = null;
		$this->count = 1;
	}
}
class StringTree {
	protected $root;
	Function __construct() {
		$this->root = NULL;
	}
	public Function Display() {
		$this->DisplayUtil ( $this->root );
	}
	protected Function DisplayUtil($curr) {
		iF ($curr != NULL) {
			echo (" value is ::" . $curr->value);
			echo (" count is :: " . $curr->count . "<br/>");
			
			$this->DisplayUtil ( $curr->lChild );
			$this->DisplayUtil ( $curr->rChild );
		}
	}
	public Function Insert($value) {
		$this->root = $this->InsertUtil ( $value, $this->root );
	}
	protected Function InsertUtil($value, $curr) {
		$compare = null;
		iF (($curr == NULL)) {
			$curr = new Node ( $value );
		} else {
			$compare = strcmp ( $curr->value, $value );
			iF (($compare == 0))
				++ $curr->count;
			else iF (($compare == 1))
				$curr->lChild = $this->InsertUtil ( $value, $curr->lChild );
			else
				$curr->rChild = $this->InsertUtil ( $value, $curr->rChild );
		}
		return $curr;
	}
	public Function FreeTree() {
		$this->root = NULL;
	}
	public Function Find($value) {
		$ret = $this->FindUtil ( $this->root, $value );
		echo ("Find " . $value . " Return " . $ret . "<br/>");
		return $ret;
	}
	protected Function FindUtil($curr, $value) {
		$compare = null;
		iF ($curr == NULL)
			return FALSE;
		$compare = strcmp ( $curr->value, $value );
		iF (($compare == 0))
			return TRUE;
		else {
			iF (($compare == 1))
				return $this->FindUtil ( $curr->lChild, $value );
			else
				return $this->FindUtil ( $curr->rChild, $value );
		}
	}
	public Function Frequency($value) {
		$ret = $this->FrequencyUtil ( $this->root, $value );
		echo ("Frequency " . $value . " Return " . $ret . "<br/>");
	}
	protected Function FrequencyUtil($curr, $value) {
		$compare = null;
		iF ($curr == NULL)
			return 0;
		$compare = strcmp ( $curr->value, $value );
		iF ($compare == 0)
			return $curr->count;
		else {
			iF ($compare > 0)
				return $this->FrequencyUtil ( $curr->lChild, $value );
			else
				return $this->FrequencyUtil ( $curr->rChild, $value );
		}
	}
}

$tt = new StringTree ();
$tt->Insert ( "banana" );
$tt->Insert ( "apple" );
$tt->Insert ( "mango" );
$tt->Insert ( "banana" );
$tt->Insert ( "apple" );
$tt->Insert ( "mango" );
echo ("Search results For apple, banana, grapes and mango :<br/>");
	
$tt->Find ( "apple" );
$tt->Find ( "banana" );
$tt->Find ( "banan" );
$tt->Find ( "napplkhjkhkju" );
$tt->Find ( "grapes" );
$tt->Find ( "mango" );
$tt->Display ();

$tt->Frequency ( "apple" );
$tt->Frequency ( "banana" );
$tt->Frequency ( "mango" );
$tt->Frequency ( "hemant" );
?>

File: Data-Structures-and-Algorithms-using-Swift/String/StringTree.swift
class StringTree {
	class Node {
		var value : String
		var count : Int
		var lChild : Node?
		var rChild : Node?

		init(_ value : String){
			self.value = value
			self.count = 0
			self.lChild = nil
			self.rChild = nil
		}

		init(){
			self.value = ""
			self.count = 0
			self.lChild = nil
			self.rChild = nil
		}
	}

	var root : Node?

	public func display() {
		self.displayUtil(curr : self.root)
	}
	
	private func displayUtil(curr : Node?) {
		guard let curr = curr else {
			return
		}
		print(" value is :: \(curr.value)")
		print(" count is :: \(curr.count)")
		self.displayUtil(curr : curr.lChild)
		self.displayUtil(curr : curr.rChild)
	}
	
	public func insert(_ value : String) {
		self.root = self.insertUtil(value : value, curr : self.root)
	}
	
	private func insertUtil(value : String, curr : Node?) -> Node {
		guard let curr = curr else {
			let temp = Node(value)
			return temp
		} 
		if curr.value == value {
			curr.count+=1
		} else if curr.value > value {
			curr.lChild = self.insertUtil(value : value, curr : curr.lChild)
		} else {
			curr.rChild = self.insertUtil(value : value, curr : curr.rChild)
		}
		return curr
	}
	
	public  func freeTree() {
		self.root = nil
	}
	
	public func find(_ value : String) -> Bool {
		let ret = self.findUtil(curr : self.root, value : value)
		return ret
	}
	
	private func findUtil(curr : Node?, value : String) -> Bool {
		guard let curr = curr else {
			return false
		}

		if curr.value == value {
			return true
		}
	
		if curr.value > value {
			return self.findUtil(curr : curr.lChild, value : value)
		}
		return self.findUtil(curr : curr.rChild, value : value)
	}
	
	public func frequency(_ value : String) -> Int {
		return self.frequencyUtil(curr : self.root, value : value)
	}
	
	private func frequencyUtil(curr : Node?, value : String) -> Int {
		guard let curr = curr else {
			return 0
		}
		if curr.value == value {
			return curr.count
		}
		if curr.value > value {
			return self.frequencyUtil(curr : curr.lChild, value : value)
		}
		return self.frequencyUtil(curr : curr.rChild, value : value)
	}

}

// Testing code.
var tt = StringTree()
tt.insert("banana")
tt.insert("apple")
tt.insert("mango")
print("Apple Found :", tt.find("apple"))
print("Banana Found :", tt.find("banana"))
print("Grapes Found :", tt.find("grapes"))
